dauphine-open-data-lib/src/main/java/io/github/oliviercailloux/y2018/opendata/dao/CourseDao.java

```diff
@@ -11,4 +18,44 @@ public CourseDao() {
 		super("Course", Course.class);
 	}
 
+	/**
+	 * Returns the list of courses that match the parameters passed by argument
+	 *
+	 * We work with the assumption that someone using our program is trying to find data with non-empty fields.
+	 * If a parameter is empty, it's not used in the SQL query.
+	 * Therefore, if all parameters are empty, no results are returned
+	 *
+	 * The search works like SQL LIKE. If a course is named "Java", it will be returned
+	 * whether the query is "Java", "Javaa" or "aJava". On the other hand, "Jva", "Jaa" or "Jvaa" won't be returned.
+	 *
+	 * @param name
+	 * @param desc
+	 * @param lang
+	 * @return
+	 */
+
+	public List<Course> findByNameDescLang(String name, String desc, String lang) {
+		if (name.isEmpty() && desc.isEmpty() && lang.isEmpty()) {
+			return Collections.emptyList();
+		}
+
+		CriteriaBuilder cb = em.getCriteriaBuilder();
+
+		CriteriaQuery<Course> q = cb.createQuery(Course.class);
+		Root<Course> c = q.from(Course.class); // similar to FROM
+
+		if (!name.isEmpty()) {
+			q.where(cb.like(c.<String>get("courseName"), "%" + name + "%"));
```

Comment:
Non, ceci vous expose à de l’injection SQL.

Reply:
Bonjour,

According to [this comment](https://stackoverflow.com/questions/15065906/does-hibernate-criteria-api-completely-protect-from-sql-injection) the "Criteria API as well as query parameters in HQL or JPQL both escape the parameters and would not execute malicious SQL."